src/lib/rank_utils.ts

```diff
@@ -22,9 +22,74 @@ interface IRankInfo {
     label: string;
 }
 
-export const MaxRank: number = 36;
+class Rating {
+    unset:boolean;
+    rating:number;
+    deviation:number;
+    volatility:number;
+    rank:number;
+    rank_label:string;
+    partial_rank:number;
+    partial_rank_label:string;
+    rank_deviation_labels:Array<string>;
+    rank_deviation:number;
+}
+
+export const MaxRank: number = 39;
+
+
+const MIN_RATING = 100;
+const MAX_RATING = 6000;
+
+export function rank_to_rating(rank:number) {
+    return 900 * Math.exp(0.032 * rank);
+}
+
+export function rating_to_rank(rating:number) {
+    return Math.log(Math.min(MAX_RATING, Math.max(MIN_RATING, rating)) / 900.0) / 0.032;
+}
+
+export function get_handicap_adjustment(rating:number, handicap:number):number {
+    return rank_to_rating(rating_to_rank(rating) + handicap) - rating;
```

Comment:
So I actually ran some numbers on this while I was trying to figure out how to best come up with a mapping from rating -> ranking, here's a breakdown showing the win % by black when given an appropriate handicap as computed by the difference in ranks:

```
26k  	     17.6 [  125]      9.1 [   22]     15.8 [   19]      0.0 [   13]      3.3 [   30]      0.0 [   12]      0.0 [    6]      0.0 [   11]      3.6 [   28]
25k  	     37.4 [ 2764]     34.4 [  346]     29.3 [  259]     23.6 [  229]     22.5 [  200]     25.9 [  166]     20.5 [  210]     19.2 [  271]     13.2 [  227]
24k  	     42.9 [ 1917]     38.5 [  286]     31.1 [  235]     30.4 [  217]     27.5 [  189]     23.9 [  184]     19.2 [  214]     17.2 [  186]     19.5 [  149]
23k  	     44.6 [ 1737]     41.7 [  271]     36.6 [  273]     32.1 [  224]     25.9 [  193]     28.3 [  184]     22.7 [  141]     14.1 [  149]     19.4 [  155]
22k  	     43.4 [ 2005]     32.7 [  352]     35.1 [  342]     26.1 [  280]     30.7 [  244]     29.9 [  197]     21.9 [  178]     23.4 [  175]     21.1 [  180]
21k  	     44.8 [ 2284]     41.8 [  423]     37.3 [  445]     33.3 [  312]     32.2 [  283]     26.5 [  230]     26.5 [  245]     26.4 [  208]     18.6 [  220]
20k  	     42.6 [ 2427]     38.4 [  657]     35.2 [  576]     34.1 [  492]     35.3 [  402]     30.9 [  356]     25.6 [  324]     22.8 [  364]     31.3 [  275]
19k  	     45.5 [ 2827]     43.6 [  910]     40.4 [  680]     38.6 [  611]     34.2 [  471]     30.8 [  465]     30.9 [  447]     32.2 [  351]     34.8 [  305]
18k  	     45.4 [ 3823]     43.0 [ 1193]     38.8 [  970]     35.2 [  772]     30.5 [  649]     30.7 [  551]     36.0 [  534]     28.5 [  355]     32.3 [  220]
17k  	     43.4 [ 4390]     38.6 [ 1730]     38.6 [ 1373]     38.5 [ 1107]     36.6 [  908]     37.0 [  829]     39.1 [  594]     41.9 [  346]     29.1 [  265]
16k  	     46.7 [ 4801]     41.1 [ 2157]     39.2 [ 1644]     41.8 [ 1368]     39.6 [ 1120]     42.2 [  818]     41.5 [  511]     35.9 [  329]     37.0 [  246]
15k  	     45.9 [ 5672]     42.9 [ 2544]     43.6 [ 2173]     41.0 [ 1765]     42.3 [ 1207]     43.2 [  708]     37.6 [  474]     31.0 [  352]     44.3 [  201]
14k  	     44.7 [ 5973]     42.9 [ 3410]     42.9 [ 2842]     43.7 [ 1906]     43.1 [ 1099]     40.1 [  676]     39.2 [  469]     40.1 [  304]     39.4 [  218]
13k  	     46.3 [ 7289]     43.5 [ 4279]     44.8 [ 2892]     45.3 [ 1619]     41.7 [  972]     40.4 [  659]     41.0 [  420]     41.0 [  310]     37.9 [  206]
12k  	     46.6 [ 7866]     46.0 [ 4796]     45.3 [ 2862]     47.9 [ 1654]     44.9 [  988]     46.9 [  614]     45.4 [  441]     43.0 [  302]     47.7 [  243]
11k  	     45.3 [ 8209]     45.4 [ 4443]     45.5 [ 2505]     46.1 [ 1446]     46.2 [  894]     46.7 [  593]     41.7 [  410]     44.6 [  307]     45.7 [  188]
10k  	     46.8 [ 8007]     46.5 [ 4167]     44.9 [ 2286]     45.9 [ 1360]     45.0 [  842]     45.0 [  545]     46.1 [  406]     46.0 [  226]     43.9 [  198]
9k   	     45.7 [ 6636]     44.7 [ 3367]     44.9 [ 1947]     45.3 [ 1158]     43.9 [  572]     42.9 [  394]     42.0 [  269]     44.3 [  262]     42.0 [  231]
8k   	     45.7 [ 5142]     46.0 [ 3102]     46.2 [ 1698]     47.0 [  895]     49.2 [  516]     45.9 [  351]     42.2 [  270]     39.6 [  288]     45.4 [  141]
7k   	     46.5 [ 4968]     46.6 [ 2955]     43.2 [ 1467]     45.2 [  746]     47.2 [  407]     42.6 [  326]     43.9 [  326]     47.7 [  172]     51.0 [   49]
6k   	     45.5 [ 4927]     46.8 [ 2357]     43.6 [ 1302]     44.9 [  686]     39.7 [  451]     38.5 [  491]     38.2 [  228]     49.4 [   83]     46.2 [   39]
5k   	     45.8 [ 3889]     44.9 [ 2026]     48.1 [ 1004]     39.9 [  666]     44.7 [  517]     49.3 [  268]     50.5 [  220]     48.4 [   95]     31.2 [   64]
4k   	     46.4 [ 3489]     44.7 [ 1683]     40.3 [  980]     42.0 [  722]     43.7 [  382]     47.5 [  162]     66.7 [  108]     50.0 [   46]     42.9 [   70]
3k   	     41.4 [ 2781]     37.8 [ 1572]     38.8 [ 1094]     40.1 [  526]     45.7 [  247]     50.8 [  185]     36.0 [   86]     56.1 [   98]     50.0 [    4]
2k   	     42.0 [ 1901]     41.3 [ 1148]     40.9 [  597]     42.8 [  421]     43.3 [  275]     36.2 [   58]     54.0 [   87]     50.0 [    4]      0.0 [    0]
1k   	     41.9 [ 1469]     43.9 [  469]     31.5 [  432]     40.5 [  264]     46.2 [   65]     54.4 [   90]     66.7 [    6]      0.0 [    0]      0.0 [    0]
1d   	     41.8 [  809]     46.0 [  554]     42.3 [  208]     48.6 [  146]     73.1 [   67]    100.0 [    1]      0.0 [    0]      0.0 [    0]      0.0 [    0]
2d   	     42.7 [  403]     38.7 [  217]     53.8 [  106]     49.0 [  104]     50.0 [   12]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]
3d   	     34.5 [  267]     46.6 [   88]     56.1 [  155]    100.0 [    3]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]
4d   	     43.4 [  152]     75.3 [   77]     80.0 [    5]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]
5d   	     77.3 [  132]     33.3 [   48]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]
6d   	     11.1 [   27]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]      0.0 [    0]

```

So the first thing these numbers tell me is that when a handicap of 1 is used appropriately, it gets us as close to a 50/50 more reliably than any other handicap difference. So I don't think computing `H-0.5` would be appropriate, seems like it should be treated as a full stone, no?

The second thing they showed me is that across the board our ranks are slightly compressed, if ranks are pushed out a bit, handicaps should be awarded a little more often, and those values should approach 50%.